posts/2020/03/pytorch-tensoriterator-internals.md

```diff
@@ -0,0 +1,236 @@
+<!--
+.. title: PyTorch TensorIterator Internals
+.. slug: pytorch-tensoriterator-internals
+.. date: 2020-03-12 22:39:56 UTC-05:00
+.. tags: 
+.. category: 
+.. link: 
+.. description: 
+.. type: text
+-->
+
+<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-generate-toc again -->
+**Table of Contents**
+
+- [Introduction](#introduction)
+- [History of TensorIterator](#history-of-tensoriterator)
+    - [TH iterators](#th-iterators)
+    - [Limitations of TH iterators](#limitations-of-th-iterators)
+- [Basics of TensorIterator](#basics-of-tensoriterator)
+- [Performing iterations](#performing-iterations)
+    - [Iteration details](#iteration-details)
+        - [Setting tensor iteration dimensions](#setting-tensor-iteration-dimensions)
+- [Conclusion](#conclusion)
+
+<!-- markdown-toc end -->
+
+# Introduction
+
+Pytorch is a massive
+codebase (approx. 12 GB after taking into account the build and generated files), and having
+a method for iterating over tensors in a very efficient manner that is independent of
+data type, dimension, striding and hardware is a critical feature that can lead to a very
+massive simplification of the codebase and make distributed development much faster and
+smoother. The [`TensorIterator`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorIterator.cpp) C++ class within pytorch is a complex yet useful class that
+is used for iterating over the elements of a tensor over any dimension and implicitly
+parallelizing various operations in a device independent manner.
+
+It does this through
+a C++ API that is independent of type and device of the tensor, freeing the programmer
+of having to worry about the datatype or device when writing iteration logic for pytorch
+tensors. For those coming from the numpy universe, `NpyIter` is a close cousin of `TensorIterator`.
+
+This post is deep dive into the working of `TensorIterator` and how it works, and is
```

Comment:
"is a deep dive"

Reply:
"into the working of `TensorIterator` and how it works," --> "into how `TensorIterator` works"